crates/c/microsimulator/microsim.c

```diff
@@ -0,0 +1,109 @@
+#include <stdio.h>
+#include <complex.h>
+#include <math.h>
+#include <string.h>
+#include <stdlib.h>
+#include "qibo_core_c.h"
+#include "microsim.h"
+
+
+typedef struct {
+    complex double h[4];
+    complex double x[4];
+    complex double y[4];
+    complex double z[4];
+} Matrices;
+
+
+Matrices create() {
+    double const h = 1.0 / sqrt(2);
+    Matrices const m = {
+        {h, h, h, -h},
+        {0, 1, 1, 0},
+        {0, -I, I, 0},
+        {1, 0, 0, -1}
+    };
+    return m;
+}
+
+int compare(const void *a, const void *b) {
+    return (*(int*)a - *(int*)b);
+}
+
+size_t control_index(size_t const g, size_t const *qubits, size_t const nqubits) {
+    size_t i = g;
+    for (size_t j = 0; j < nqubits; j++) {
+        size_t const n = qubits[j];
+        size_t const k = 1 << n;
+        i = ((i >> n) << (n + 1)) + (i & (k - 1)) + k;
```

Comment:
> E.g. `i = 1010_1101` and `n = 5` results in:
> 
> * `(i >> n) << (n + 1) = 0100_0000`

I believe there is an error here, the correct result would be `10100_0000` (the first `1` is not removed). That would make the final result `10110_1101`, which is the original `i` with a `1` added in its `n=5`th position (counting from the right).

An equivalent Python function that is more readable would be
```py
def control_index_py(g, qubits, n):
    # qubits needs to be sorted!
    bits = [int(b) for b in format(g, f"0{n}b")]
    for q in qubits:
        bits.insert(q, 1)
    # convert bits to integer
    return (2 ** np.arange(len(bits) - 1, -1, -1)).dot(bits)
```
In words, `control_index` accepts an integer `g < 2 ** (nqubits - len(qubits))` and inserts 1s in its binary representation, in the places dictated by `qubits`. Thus the output is a number `< 2 ** nqubits`, or if you think in terms of arrays of booleans/bits, the output is `len(qubits)` longer than the input. Note that `qubits` here are all the elements involved in the gate (in our case that is arbitrarily many controls + one target).

I was also not sure, so I confirmed this by asserting the result of the two functions is the same for different (random) configurations of `qubits`.
<details>
<summary>Code for the check here</summary>

```py
import numpy as np

def control_index(g: int, qubits: list[int]):
    i = g
    for n in qubits:
        k = 1 << n
        i = ((i >> n) << (n + 1)) + (i & (k - 1)) + k
    return i


def control_index_py(g, qubits, n):
    bits = [int(b) for b in format(g, f"0{n}b")]
    for q in qubits:
        bits.insert(q, 1)
    # convert bits to integer
    return (2 ** np.arange(len(bits) - 1, -1, -1)).dot(bits)


nqubits = 12
a = np.arange(nqubits)

for ncontrols in range(nqubits - 1):    
    # for various random choices
    for _ in range(100):
        # grab `ncontrols` + 1 target qubits
        np.random.shuffle(a)
        qubits = a[:ncontrols + 1]
        # compare the two functions
        nopen = nqubits - len(qubits)
        for g in range(2 ** nopen):
            result1 = control_index(g, sorted(nqubits - q - 1 for q in qubits))
            result2 = control_index_py(g, sorted(qubits), nopen)
            assert result1 == result2
```

</details>

Regarding why this is needed: In order to apply a single-qubit (not controlled) gate with matrix `G[i, j]` to the state `|b1 b2 ... bN>` (where `bi` are bits) we need to do the following:
```
|b1 b2 ... 0 ... bN> --> G[0, 0] |b1 b2 ... 0 ... bN> + G[0, 1] |b1 b2 ... 1 ... bN>
|b1 b2 ... 1 ... bN> --> G[1, 0] |b1 b2 ... 0 ... bN> + G[1, 1] |b1 b2 ... 1 ... bN>
```
for all 2^(N-1) bitstrings (excluding the target). These are lines 67-68 in the code.

If the gate is controlled, the bitstrings for which ALL controls are 1 follow the above rule, while for all other bistrings we do nothing (since we update the state in place, otherwise we would have to copy). The bitstring for which all controls are 1 are 2^(N - Ncontrols - 1) and the way we generate them is to loop `g` in `range(2 ** (nqubits - ncontrols - 1))` in https://github.com/qiboteam/qibo-core/blob/5dd4ba808dd5090d897f62789805fb1e7575222d/crates/c/microsimulator/microsim.c#L62

and use `control_index` to insert 1s in all control (and the target) positions. The target is also flipped to 0 in https://github.com/qiboteam/qibo-core/blob/5dd4ba808dd5090d897f62789805fb1e7575222d/crates/c/microsimulator/microsim.c#L64

to apply the equation above.

If that explanation makes sense, it could be useful to introduce part of it as documentation in qibojit. I agree that this part is not well documented and even the variable names are not good. Historically, we first introduced this in C++ (and also CUDA) for qibotf, then moved to numba (Python) for qibojit. I also kind of invented it myself, without looking what other libraries are doing, so maybe is not even the best implementation, but I guess we kept it because it was working (at least noone complained so far) and performance is acceptable.